app-experimental.tex

```diff
@@ -2693,6 +2693,125 @@ \section{Distributed Capabilities For Peripherals And Accelerators} % <<<
+For example, in the library compartmentalisation case, some malicious compartment could stash the current capability of TDC in some memory. Later, the malicious compartment could overwrite TDC with stale value and thus confuse trusted code.
+
+\subsubsection{Nestability of TDC}
+
+In the normal case, the type of the value in TDC is the one expected by the current level of compartmentalisation.
+However, when trusted code encounters a data capability that is not of its own type, it must have been installed because a nested compartmentalisation mechanism uses a subtype.
+The trampoline can use its thread ID to access the correct TDC.
+It could also call more privileged code, but that would result in great performance penalty.
+Once the trampoline is ready to return, it will need to re-instantiate the trusted data capability it has swapped out when being entered.
+
+\subsubsection{Usage of TID register}
+
+In the code below, we present how we expect the TID to be used in a system in order to load a trusted data capability.
+The respective TID register is used to index into a table of trusted stacks.
+In order to ensure correctness, the current data capability needs to be updated in a part of memory that is accessible by all trusted code.
+This approach does not require the sentry to have the write permission set.
+The following code sequence is needed to load the TDC:
+
+\begin{lstlisting}[label=tid_load_tdc]
+auipcc ct0, const_0
+CIncOffsetImm ct0, ct0, const_1
+clc ct0, 0(ct0) // load cap to sealed trusted caps
+csrr ct1, tid
+CIncOffset ct0, ct0, t1
+clc ct0, 0(ct0) // use tid as offset
+\end{lstlisting}
+
+\subsubsection{Usage of TDC register}
+
+With the TDC implemented, one only needs to load the authorising capability relative to the PCC.
+While everyone can read the TDC register, its content can only be unsealed with a suitable authorising capability.
+The code below shows how a trampoline can retrieve its trusted data capability:
+
+\begin{lstlisting}[label=tdc_load_tdc]
+auipcc ct0, const_0
+CIncOffsetImm ct0, ct0, const_1
+clc ct0, 0(ct0) // load auth cap
+csrr ct1, tdc // everyone can read TDC
+CUnseal ct0, ct0, ct1 // unseal TID
+\end{lstlisting}
+
+\subsection{Implementation Notes}
+
+Access to CSR registers is usually slow in microarchitectures because many CSRs take effect on the system and thus need to stall the pipeline or even need to redirect.
+The TID and TDC registers are different because they are pure-data registers and no value stored in it has any effect on other architectural state.
+Thus, we propose to treat these registers differently.
+
+We note that the TID register is written when every there is a thread switch, but not otherwise.
+However, the TID register is read on every trampoline entry.
+Thus, we conclude that the reading speed of the TID register matters for overall performance whereas the writing speed does not.
+We envision that this strategy might be useful to other pure-data CSRs.
```

Comment:
Perhaps it would be useful to better characterise / discuss the performance considerations here. In the CheriBSD library compartmentalisation model, you’d need to context switch TID every time you context switch a user thread, but not every time you perform a domain switch within a user thread. Thread context switching speeds do matter, so we need to be cautious in the above statement: it is less performance critical, but it’s not irrelevant.